src/main/java/io/confluent/idesidecar/restapi/kafkarest/controllers/ClusterManagerImpl.java

```diff
@@ -0,0 +1,144 @@
+import io.smallrye.mutiny.Uni;
+import io.smallrye.mutiny.infrastructure.Infrastructure;
+import io.vertx.core.http.HttpServerRequest;
+import jakarta.enterprise.context.RequestScoped;
+import jakarta.inject.Inject;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Supplier;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.DescribeClusterResult;
+import org.apache.kafka.common.Node;
+
+/**
+ * RequestScoped bean for managing Kafka clusters. Creating the bean as {@link RequestScoped} allows
+ * us to inject the {@link HttpServerRequest} which is used to get the connection ID from the
+ * request headers.
+ */
+@RequestScoped
+public class ClusterManagerImpl implements ClusterManager {
+
+  @Inject
+  AdminClients adminClients;
+
+  @Inject
+  HttpServerRequest request;
+
+  Supplier<String> connectionId = () -> request.getHeader(CONNECTION_ID_HEADER);
+
+  @Override
+  public Uni<ClusterData> getKafkaCluster(String clusterId) {
+    return describeCluster(clusterId)
+        .chain(cid -> {
+          if (!cid.id().equals(clusterId)) {
+            return Uni.createFrom().failure(new ClusterNotFoundException(
+                "Kafka cluster '%s' not found.".formatted(clusterId)
+            ));
+          }
+          return uniItem(cid);
+        })
+        .map(this::fromClusterId);
+  }
+
+  @Override
+  public Uni<ClusterDataList> listKafkaClusters() {
+    return describeCluster(null)
+        .chain(cluster -> uniItem(ClusterDataList
+            .builder()
+            .metadata(ResourceCollectionMetadata
+                .builder()
+                .self(forClusters().getRelated())
+                // We don't support pagination
+                .next(null)
+                .build()
+            )
+            .kind("KafkaClusterList")
+            .data(List.of(fromClusterId(cluster)))
+            .build())
+        );
+  }
+
+  private Uni<ClusterDescribe> describeCluster(String clusterId) {
+    return uniItem((Supplier<AdminClient>)
+        () -> adminClients.getAdminClient(connectionId.get(), clusterId))
+        .runSubscriptionOn(Infrastructure.getDefaultWorkerPool())
+        .map(supplier -> supplier.get().describeCluster())
+        .chain(describeClusterResult ->
+            uniStage(describeClusterResult.clusterId().toCompletionStage())
+                .map(id -> new ClusterDescribe(describeClusterResult).withId(id)))
+        .chain(cid -> uniStage(cid.result.controller().toCompletionStage()).map(Node::id)
+            .map(cid::withControllerId)
+        ).chain(cid -> uniStage(cid.result.nodes().toCompletionStage())
+            .map(cid::withNodes)
+        );
+  }
+
+  private ClusterData fromClusterId(ClusterDescribe cluster) {
+    return ClusterData
+        .builder()
+        .kind("KafkaCluster")
+        .metadata(ResourceMetadata
+            .builder()
+            .self(getClusterCrn(cluster.id()).toString())
```

Comment:
Should we include CRN if this is not CCloud? If it is CCloud, then can we not use the existing `Crn` factory methods?

Reply:
No, I've made `resource_name` null. Also, `self` here should've been `RelationshipUtil.forCluster(clusterId)` and not a CRN. I've fixed that.